### ISSUE TITLE
Buggy results when operating on array copied with astype() (migrated from Trac #1863)

### ISSUE THREAD

**Issue by thouis** [thouis/numpy-trac-migration#3411]
Original ticket http://projects.scipy.org/numpy/ticket/1863
Reported 2011-06-08 by trac user embray, assigned to unknown.

I'm getting really buggy results from any mathematical operation I do on an array of float64 that was obtained by calling `arr.astype('float64')` on an existing array of float64s.

Here's an example that illustrates the problem (the fft stuff is irrelevant; it just creates an array similar to the one my code is bugging out on):

```
#!python
import numpy
from numpy import fft


shape = (32, 16)
type = numpy.complex64
true = numpy.arange(shape[0] * shape[1], dtype=type)
true.shape = shape

y = fft.fft(true, shape[0], 0)
y = y.imag

z = y.astype(numpy.float64)

print (y ** 2)[3]
print (z ** 2)[3]
```

This particular example also throws a `RuntimeWarning: overflow encountered in square` when squaring `z`, the copy of the original array `y`.

This just started happening recently using the latest from git.  I'm betting it has something to do with [https://github.com/numpy/numpy/commit/a17a4996e4ed63d1b855a0917fb5fcdd5855a7d0 this commit] that made changes to astype(), but I haven't tracked down the exact issue yet.

**Comment by thouis** [comment]
Comment in Trac by atmention:mwiebe, 2011-06-09

This bug turns out to be much more longstanding than is implied by it showing up in astype. It's a bit trickier, but possible to expose it in 1.6 as well:

```
>>> np.__version__
'1.6.1.dev-878ab94'
>>> a = np.arange(24).reshape(2,3,4).swapaxes(0,1)
>>> a.strides
(16, 48, 4)
>>> b = np.zeros_like(a)
>>> b.strides
(16, 48, 4)
>>> a.flags.c_contiguous
False
>>> b.flags.c_contiguous
True
```

The C_CONTIGUOUS flag should not be set in this case, and that's what was happening in the bug report example too, just with a simpler array and different function call.

Fixed in commit:a1e7be3 (master) and commit:5765325 (1.6.1).

**Comment by thouis** [comment]
Comment in Trac by trac user embray, 2011-06-09

I see now--thanks for tracking that down so quickly!
